File: src/deps/glaml_extended/yaml_ffi.mjs
import * as fs from "node:fs";
import yaml from "js-yaml";
import { Ok, Error, toList } from "../../../prelude.mjs";
import {
  NodeStr,
  NodeInt,
  NodeFloat,
  NodeBool,
  NodeNull,
  NodeSeq,
  NodeMap,
} from "./yaml.mjs";

// Parse YAML file and return list of documents
export function parse_file(path) {
  try {
    const content = fs.readFileSync(path, "utf8");
    const docs = yaml.loadAll(content);
    // Wrap each doc for document_root to unwrap
    const gleamDocs = docs.map((doc) => ({ _yaml_doc: doc }));
    return new Ok(toList(gleamDocs));
  } catch (e) {
    return new Error(e.message || "Failed to parse YAML");
  }
}

// Parse YAML string and return list of documents
export function parse_string(content) {
  try {
    const docs = yaml.loadAll(content);
    const gleamDocs = docs.map((doc) => ({ _yaml_doc: doc }));
    return new Ok(toList(gleamDocs));
  } catch (e) {
    return new Error(e.message || "Failed to parse YAML");
  }
}

// Get root node from document wrapper
export function document_root(doc) {
  return jsToNode(doc._yaml_doc);
}

// Convert JS value to Gleam Node type
function jsToNode(value) {
  if (value === null || value === undefined) {
    return new NodeNull();
  }
  if (typeof value === "string") {
    return new NodeStr(value);
  }
  if (typeof value === "number") {
    if (Number.isInteger(value)) {
      return new NodeInt(value);
    }
    return new NodeFloat(value);
  }
  if (typeof value === "boolean") {
    return new NodeBool(value);
  }
  if (Array.isArray(value)) {
    return new NodeSeq(toList(value.map(jsToNode)));
  }
  if (typeof value === "object") {
    const entries = Object.entries(value).map(([k, v]) => [
      jsToNode(k),
      jsToNode(v),
    ]);
    return new NodeMap(toList(entries));
  }
  return new NodeNull();
}

// Parse integer from string
export function int_parse(s) {
  const n = parseInt(s, 10);
  if (isNaN(n)) {
    return new Error(undefined);
  }
  return new Ok(n);
}

// Pop first grapheme from string
export function pop_grapheme(s) {
  if (s.length === 0) {
    return new Error(undefined);
  }
  // Handle Unicode properly
  const chars = [...s];
  return [chars[0], chars.slice(1).join("")];
}

File: src/deps/glaml_extended/yaml_ffi.erl
-module(yaml_ffi).
-export([parse_file/1, parse_string/1, document_root/1, int_parse/1]).

-include_lib("yamerl/include/yamerl_errors.hrl").

%% Ensure yamerl is started (cached after first call)
-define(ENSURE_YAMERL,
    case whereis(yamerl_app) of
        undefined -> application:ensure_all_started(yamerl);
        _ -> ok
    end).

%% Parse YAML file
parse_file(Path) ->
    ?ENSURE_YAMERL,
    try
        Docs = map_yamerl_docs(yamerl_constr:file(Path, [{detailed_constr, true}])),
        {ok, Docs}
    catch
        throw:#yamerl_exception{errors = [First | _]} ->
            {error, format_error(First)};
        error:Reason ->
            {error, iolist_to_binary(io_lib:format("~p", [Reason]))}
    end.

%% Parse YAML string
parse_string(Content) ->
    ?ENSURE_YAMERL,
    try
        Docs = map_yamerl_docs(yamerl_constr:string(Content, [{detailed_constr, true}])),
        {ok, Docs}
    catch
        throw:#yamerl_exception{errors = [First | _]} ->
            {error, format_error(First)};
        error:Reason ->
            {error, iolist_to_binary(io_lib:format("~p", [Reason]))}
    end.

%% Get document root
document_root(Doc) ->
    {document, Root} = Doc,
    Root.

int_parse(S) ->
    try
        {ok, binary_to_integer(S)}
    catch
        _:_ -> {error, nil}
    end.

%% Format yamerl error to string
format_error(#yamerl_parsing_error{text = undefined}) ->
    <<"Unexpected parsing error">>;
format_error(#yamerl_parsing_error{text = Message}) ->
    unicode:characters_to_binary(Message);
format_error(_) ->
    <<"Unknown YAML error">>.

%% Convert yamerl documents to our format
map_yamerl_docs(Documents) ->
    lists:map(fun map_yamerl_doc/1, Documents).

map_yamerl_doc({yamerl_doc, RootNode}) ->
    {document, map_yamerl_node(RootNode)}.

%% Convert yamerl nodes to yaml.gleam compatible format
map_yamerl_node({yamerl_null, _, _Tag, _Loc}) ->
    node_null;
map_yamerl_node({yamerl_str, _, _Tag, _Loc, String}) ->
    {node_str, unicode:characters_to_binary(String)};
map_yamerl_node({yamerl_bool, _, _Tag, _Loc, Bool}) ->
    {node_bool, Bool};
map_yamerl_node({yamerl_int, _, _Tag, _Loc, Int}) ->
    {node_int, Int};
map_yamerl_node({yamerl_float, _, _Tag, _Loc, Float}) ->
    {node_float, Float};
map_yamerl_node({yamerl_seq, _, _Tag, _Loc, Nodes, _Count}) ->
    {node_seq, lists:map(fun map_yamerl_node/1, Nodes)};
map_yamerl_node({yamerl_map, _, _Tag, _Loc, Pairs}) ->
    {node_map, [{map_yamerl_node(K), map_yamerl_node(V)} || {K, V} <- Pairs]}.
